### DISCUSSION TITLE
Forwarding the contents of Callback and URL buttons

### DISCUSSION THREAD

**Discussion by ivanivanoff888** [aahnik/tgcf#224] [Q&A] [answered]
Hence is the question, how to forward the contexts of Callback (text) and URL buttons (URL-links)?

![4fad2e2743dc8eda04](https://user-images.githubusercontent.com/86550396/123553076-bcf06400-d792-11eb-9554-9d9f6e0267b3.png)
![64d0d842ff73ff3dc8](https://user-images.githubusercontent.com/86550396/123553077-be219100-d792-11eb-99c1-4979e8b24eaf.png)

**Comment by aahnik** [comment, marked as answer]
This has already been discussed
https://github.com/aahnik/tgcf/issues/209#issuecomment-868952068

**Reply by ivanivanoff888** [reply]
Well, that issue was mostly about inline buttons (or it is mixed up with callback buttons) (There is a channel with inline buttons. If you click on the button, a notification appears with information. I would like to forward information from a notification or the information button itself). The aim is not to get the info that is received after clicking the button. The aim is to get the TEXT OF THE BUTTON.
Also if you change several settings in css-file in Telegram web-version this text becomes available to copying.

"Buttons that contain URL (clicking on these buttons is like clicking on a link)
When you forward a message by hand, the buttons of type 1 are never forwarded. It is not possible to forward them. But buttons of type 2 are forwarded." - By default callback and URL buttons are lost during forwarding.

"Tgcf also forwards buttons of type 2. You must set show_forwarded_from: true to enjoy this feature. Otherwise, URL buttons will not get forwarded." - That is a good news. If URL-buttons can be forwarded, then the text of the callback buttons also can be forwarded.

Info about types of buttons: https://core.telegram.org/bots/2-0-intro.

**Reply by aahnik** [reply]
Hi @ivanivanoff888 please try to understand my point. Read this https://github.com/aahnik/tgcf/issues/209#issuecomment-868952068 again.

> that issue was mostly about inline buttons 

Your question is also about inline buttons. There are two kinds of them. One that contain URLs, and others that contain data.

> If you click on the button, a notification appears with information.

That response is delivered by a bot. When a button with callback data is clicked, the bot running in that channel/group/direct chat receives an update with some data, as set by the developer. The bot responds with a notification text. If that particular bot is not running, you will not get a response.

> css-file in Telegram web-version

This project does not deal with that. Dont go offtopic. This project is built with Telethon, and connects with Telegram's Servers directly via mtproto protocol. We don't use any web scrapping here.

>  That is a good news. If URL-buttons can be forwarded, then the text of the callback buttons also can be forwarded.

No dear. It does not work that way. The other type of inline buttons is not forwarded. Those are constructed by a bot. Try to forward such a message by hand in the Telegram app. You will see such buttons get lost after forwarding. If a message has URL buttons, they are preserved. Just try forwarding by hand.

Tgcf can be used by either a user account or a bot account. A user account cant create buttons. But a bot account can. Although a bot **cannot forward inline buttons with callback data, it can construct similar buttons and send a new message object.**

But what would be the benefit of that? Because clicking on these new buttons won't do anything.

If you seriously need the feature you can write your own tgcf plugin, or contact me on [Telegram](https://aahnik.dev).
If you are contacting me on Telegram, please explain what you are actually trying to do. Dont fall in the [xy problem trap](https://xyproblem.info/).
